### PR TITLE
Check if pooled connection is still open

### PR DESCRIPTION
**Pull request by arhimondr**
Closing a netty channel is an asynchronous operation. Before the connection
is closed, a new request may come in and grab a connection that is about to be
closed by the previous request.

### CODE REVIEW

File: drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/ThriftServerHandler.java [hunk 1]
```diff
@@ -92,6 +98,17 @@ public void channelRead(ChannelHandlerContext context, Object message)
         context.fireChannelRead(message);
     }
 
+    @Override
+    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause)
```
Comment (dain) [inline drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/ThriftServerHandler.java]: Is this the right place to put this?  The ThriftServerHandler is the inner most handler on the server.  Does it get exceptions from the ChannelHandlers closer to the server socket?

File: drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/ThriftServerHandler.java [hunk 2]
```diff
@@ -107,6 +107,10 @@ public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause)
         }
 
         log.error(cause);
+
+        if (ctx.channel().isActive()) {
```
Comment (dain) [inline drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/ThriftServerHandler.java]: Do we need the `if` check?  Why not unconditionally close the context.

File: drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/ThriftServerHandler.java (hunk 1 above)
Comment (arhimondr) [reply drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/ThriftServerHandler.java]: I think the chain is next

`LengthFieldBasedFrameDecoder` -> `SimpleFrameCodec` -> `ResponseOrderingHandler` -> `ThriftServerHandler`

If something fails in upstream handlers, the error is propagated all the way downstream to the `ThriftServerHandler` (unless the error is handled somewhere upstream. `SimpleFrameCodec` and `ResponseOrderingHandler` don't have the  `exceptionCaught` method, so the error is propagated all the way to the `ThriftServerHandler`.

Comment (arhimondr) [reply drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/ThriftServerHandler.java]: I'm not sure how the write path works though. What is the propagation chain if the error happens when response is written to the client.

File: drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/ThriftServerHandler.java (hunk 2 above)
Comment (arhimondr) [reply drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/ThriftServerHandler.java]: I took it from the `io.netty.handler.ssl.SslHandler`. I'm not sure why it was added there. Maybe just to safe some cycles on Future creation.

**Comment by arhimondr** [comment]
@dain @electrum comments addressed

File: drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/ThriftServerHandler.java (hunk 1 above)
Comment (dain) [reply drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/ThriftServerHandler.java]: I dug into this a bit more.  Uncaught exceptions are propagated down the `ChannelInboundHandler` path.  This handler is the end of this path so this should work.

File: drift-integration-tests/src/test/java/io/airlift/drift/integration/guice/TestGuiceIntegration.java [hunk 3]
```diff
@@ -211,6 +211,20 @@ private static void assertEchoService(EchoService service)
 
     private static void assertThrowingService(ThrowingService service)
     {
+        try {
```
[the remark this answers is not in the document]
Comment (arhimondr) [reply drift-integration-tests/src/test/java/io/airlift/drift/integration/guice/TestGuiceIntegration.java]: Yes. So if does more requests after it fails with `TooLargeFrame`. That's the way i caught the bug.

File: drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/client/ConnectionPool.java [hunk 4]
```diff
@@ -69,26 +69,35 @@ public ConnectionPool(ConnectionManager connectionFactory, EventLoopGroup group,
...
+                    throw new RuntimeException(e);
+                }
+
+                // connection is still opening
+                if (!future.isDone()) {
+                    return future;
+                }
+
+                // check if connection is failed or closed
+                Channel channel = future.getNow();
```
Comment (electrum) [inline drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/client/ConnectionPool.java]: Inline this variable?

**Pull request merged**